src/Simulation/Simulators/OpenSystemsSimulator/DataModel/StateConverter.cs

```diff
@@ -1,112 +1,100 @@
-                        }
-                    );
+    public override State Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        reader.Require(JsonTokenType.StartObject, readCurrent: false);
 
-                    if (value is ArrayState { Data: var data })
+        var arrayConverter = new ComplexArrayConverter();
+        return reader.ReadQubitSizedData<State>((ref Utf8JsonReader reader, string kind) =>
+            kind switch
+            {
+                "Pure" => arrayConverter.Read(ref reader, typeof(NDArray), options).Bind(
+                    (int nQubits, NDArray data) => new PureState(nQubits, data)
+                ),
+                "Mixed" => arrayConverter.Read(ref reader, typeof(NDArray), options).Bind(
+                    (int nQubits, NDArray data) => new MixedState(nQubits, data)
+                ),
+                "Stabilizer" => JsonSerializer.Deserialize<StabilizerState>(ref reader).Bind(
+                    (int nQubits, StabilizerState? state) =>
                     {
-                        arrayConverter.Write(writer, data, options);
+                        System.Diagnostics.Debug.Assert((state?.Data as object) != null);
+                        System.Diagnostics.Debug.Assert(nQubits == state.NQubits);
+                        return state;
                     }
-                    else if (value is StabilizerState stabilizerState)
+                ),
+                _ => throw new JsonException($"Unknown state kind {kind}.")
+            }
+        );
+    }
+
+    public override void Write(Utf8JsonWriter writer, State value, JsonSerializerOptions options)
+    {
+        var arrayConverter = new ComplexArrayConverter();
+
+        writer.WriteStartObject();
+            writer.WriteNumber("n_qubits", value.NQubits);
+
+            writer.WritePropertyName("data");
+            writer.WriteStartObject();
+                writer.WritePropertyName(
+                    value switch
                     {
-                        var array = new StabilizerState.TableArray
-                        {
-                            Data = stabilizerState.Data.flat.ToArray<bool>().ToList(),
-                            Dimensions = stabilizerState.Data.Shape.Dimensions.ToList()
-                        };
-                        writer.WriteStartObject();
-                            writer.WritePropertyName("n_qubits");
-                            writer.WriteNumberValue(stabilizerState.NQubits);
-                            writer.WritePropertyName("table");
-                            JsonSerializer.Serialize(writer, array);
-                        writer.WriteEndObject();
+                        PureState _ => "Pure",
+                        MixedState _ => "Mixed",
+                        StabilizerState _ => "Stabilizer",
+                        _ => throw new JsonException($"Unknown state type {value.GetType()}.")
```

Comment:
(Paranoid, up to you)
During writing the `writer.Flush()` immediately before `throw new JsonException()` can probably simplify the failure debugging/analysis.